src/main/java/eu/europeana/metis/sandbox/service/workflow/ExternalValidationService.java

```diff
@@ -12,6 +13,7 @@
    * @return ordered record that is valid against EDM-EXTERNAL schema
    * @throws NullPointerException      if record is null
    * @throws RecordProcessingException if records fails at ordering
+   * @throws RecordValidationException if records fails at validation
```

Comment:
Should these exceptions ultimately inherit from RuntimeException? That sometimes makes it less obvious what is being thrown and what the calling methods should account for. Would it be a good idea to convert them to inherit from Exception instead?

Reply:
Probably you already notice all exceptions I created inherit from `RuntimeException`, I think there is debate on the use of checked exceptions and based on my experience I lean towards using unchecked exceptions but documenting them in the interface contract so that whoever uses it will know what to account for (as Joshua Bloch recommends to document every exception your method can throw) and handle or not if they want. I am not saying my approach is right or wrong, I had this talk on the past with other colleagues and listen and share thoughts on the subject and most of the time I go for an unchecked exceptions hierarchy, and handle them as convenient. But also I took some time to read again about the arguments and collected some interesting links on the topic that you can check at the bottom of this comment.

From my experience with checked exceptions I notice most of the time there is not much you can do when they happen, you mostly re-throw or catch them, and when you do catch them is to log or to wrap them into a higher level exception to re-throw again, most of the time you let them go up until they are logged, and the difference between checked and unchecked becomes that you have to declare them in your method declaration or not. So the approach that I take in general is to create an unchecked exceptions hierarchy from the project, then use those in the service level (wrapping other low level exceptions with the service ones) and document the Interface of each service to declare every exception it throws, then for example in Sandbox there are 2 clients/users of the service layer, the controller and consumer layers, which know and understand the expected exceptions from the documentation in the services interface and they handle or let them go up as they think is appropriate.

Again not saying is right or wrong, just my own experience, and since is an interesting topic for me I took time to collect info and try to make a point (and for sure there could be a lot of counter arguments). 
I just wanted to share some thoughts on the subject, but also I think there could be good reasons to use checked exceptions like most [recommendations](https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html) states 

> If a client can reasonably be expected to recover from an exception, make it a checked exception

But for me to find that point is what is not that clear, like in this case `RecordValidationException` is thrown because the record itself is not valid, so how will the user of the API (the consumer layer) can recover from that. Also quoting [Effective Java](https://www.oreilly.com/library/view/effective-java-3rd/9780134686097/)

> This burden may be justified if the exceptional condition cannot be prevented by proper use of the API and the programmer using the API can take some useful action once confronted with the exception. Unless both of these conditions are met, an unchecked exception is appropriate

Which is like most of the time for me. But I am not against and for sure I will change it if the standards you have advocate for their use, I just wanted to share my own experience and see what do you think about the topic

Here are some links for the thoughts on the topic:

- https://testing.googleblog.com/2009/09/checked-exceptions-i-love-you-but-you.html
- https://www.javaworld.com/article/3142626/are-checked-exceptions-good-or-bad.html
- https://stackoverflow.com/questions/6115896/understanding-checked-vs-unchecked-exceptions-in-java
- https://stackoverflow.com/questions/4609870/why-hibernate-changed-hibernateexception-to-unchecked-runtimeexception
- https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html
- https://stackoverflow.com/questions/613954/the-case-against-checked-exceptions
- http://literatejava.com/exceptions/checked-exceptions-javas-biggest-mistake/
- https://www.artima.com/intv/solid.html
- http://tutorials.jenkov.com/java-exception-handling/checked-or-unchecked-exceptions.html
- https://stackoverflow.com/questions/23178952/why-spring-handles-only-unchecked-exceptions/23179413

Reply:
I will admit I have not checked out all the references you gave (yet... I may do so later). But thanks for writing all this, it has focused my thinking a bit.

I think that we are in agreement on an abstract level, meaning that unchecked exceptions should be reserved for 'stuff we can't do anything about' (i.e. we can't or shouldn't be expected to recover from). To me that sometimes translates into whether we can expect an exception to occur in the normal flow of things, or whether the user (whoever that is) should see a 'nice' message or just a technical one.

The example you gave is in fact an interesting one. The RecordValidationException currently only occurs when you are checking record validity. This is in my opinion something that the tool should expect to happen (input records are often invalid, and is in fact something that people are going to use the sandbox for, to check this). So in line with our shared understanding, I would argue that it SHOULD be a checked exception, forcing code higher-up to deal with it properly and providing a meaningful feedback to the user.

However, perhaps in a different context (e.g. transformation) we know we have validated the record and the RecordValidationException should no longer occur, if it does there is a software problem that the tool is NOT expected to recover from. So perhaps there it should be thrown as or encapsulated in an unchecked/runtime exception.

So in short, I think we should spend some time identifying which exceptions are 'expected' (i.e. the user's or input's 'fault') and which aren't (i.e. the system's or programmer's fault). And make those in the former category checked? What do you think?

Reply:
Sounds good, but I think the evaluation itself deserves its own story, I can create a ticket in the backlog to be tackled on its own, for us to define the Exceptions hierarchy and explore which ones are checked and which not. (Also is the first time I think of them in terms of the kind of message that you can retrieve to the user, that is a interesting point)

Reply:
Created [MET-2622](https://europeana.atlassian.net/browse/MET-2622)